bz18870: disconnect TableView from TableModel when we're done with them.

bendk opened the pull request:
Before this was happening:
- User selects a tab, we create a table model and table view connected to
  that model.
- User moves the tab into/out of the folder
  - We re-use the table model and make a new table view
  - The old table view still would still handle the signals from the model,
    but we weren't really prepared to handle them and would throw an
    exception.

Fixed this by adding code to explicitly make a table view stop using a model.
It would be nice to support set_model() and make setting it to None a special
case, but it's not easy to do on OSX, so let's wait for this.

Commit:
bz18870: disconnect TableView from TableModel when we're done with them.

Before this was happening:
  - User selects a tab, we create a table model and table view connected to
    that model.
  - User moves the tab into/out of the folder
    - We re-use the table model and make a new table view
    - The old table view still would still handle the signals from the model,
      but we weren't really prepared to handle them and would throw an
      exception.

Fixed this by adding code to explicitly make a table view stop using a model.
It would be nice to support set_model() and make setting it to None a special
case, but it's not easy to do on OSX, so let's wait for this.

paulswartz commented on tv/osx/plat/frontends/widgets/tableview.py:
```diff
@@ -1283,15 +1283,22 @@ def __init__(self, model, custom_headers=False):
-        self.model.connect_weak('row-changed', self.on_row_change)
-        self.model.connect_weak('structure-will-change',
-                self.on_model_structure_change)
+        self.model_signal_ids = [
```
Earlier in `__init__`, there's this line:

``` python
self.data_source.initWithModel_(self.model)
```

Should we be un-initing the data source as well?

bendk replied:
Good point.  I think the best way to clear this one out is to un-set the data source for the NSTableView.  I'll add a changeset for this one.

Commit:
bz18870: alternate fix for 17823

Looks like the root cause was the changeset 33df4b57.  That one made it so
that we checked the selection when we got the structure-will-change signal.

TreeTableModel.remove() had a bug where it would:
 - emit structure-will-change
 - forget the iter
 - call the base class remove() which will
   - emit structure-will-change again
   - remove the node from the list

_check_selection() was depending on that behavior because it would catch the
WidgetActionError during the second structure-will-change signal.

Things could crash when we called remove() twice.  In that case, we haven't
called TableView.reloadData() yet, so the first row removed could still be
selected.  This time when we emit structure-will-change the node will actually
be removed, and this is what causes the crash.

This changeset:
  - Fixes TreeTableModel.remove()
  - Adds an alternate fix for #17823

Commit:
Fixup based on paul's suggestion.

Unset the datasource in TableView.unset_model()

paulswartz commented:
I think your changes look good, and 17823 is still fixed with your new code.  Merging.
